knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt

```diff
@@ -0,0 +1,131 @@
+package com.cesar.knot_sdk
+
+import com.rabbitmq.client.AMQP
+import com.rabbitmq.client.Channel
+import com.rabbitmq.client.Connection
+import com.rabbitmq.client.ConnectionFactory
+
+
+/**
+ * This class is an abstraction over the rabbitMQ java library it handles all
+ * AMQP specific operations, such as creation of connection, queues and
+ * consumers.
+ * This class contains all KNoT-Protocol specific keywords, such as, the binding
+ * keys and queue/consumer names.
+ * @param username the rabbitMQ username in the connector
+ * @param password the rabbitMQ password in the connector
+ * @param hostname the ip of the server running the connector
+ * @param port the port that rabbitMQ will be listening to (by default 5672)
+ */
+class KNoTAMQP(username : String, password : String, hostname : String, port : Int) {
+
+    val EXCHANGE_TYPE_TOPIC = "topic"
+
+    val EXCHANGE_NAME_CLOUD = "cloud"
+    val EXCHANGE_NAME_FOG = "fog"
+
+    val BINDING_KEY_REGISTER = "device.register"
+
+    val QUEUE_NAME_FOG_IN = "fogIn"
+    val QUEUE_NAME_FOG_OUT = "fogOut"
+
+    val factory = ConnectionFactory()
+
+    lateinit var conn : Connection
+    lateinit var channel : Channel
+
+    init {
+        factory.host = hostname
+        factory.username = username
+        factory.password = password
+        factory.port = port
+    }
+
+    /**
+     * Starts the connection with the rabbitMQ server. Uses the parameters
+     * defined in factory. Should not be executed in the main thread as it
```

Comment:
then shouldn't it be a suspended function? Or an async function?

Reply:
The class KNoTAMQPFactory takes care of creating a valid AMQP instance, that includes doing network operations in an async manner. For that I used "doAsyc" from the anko library. This method was preferred over putting the "doAsync" inside the KNoTAMQP class as certain operations are required to be ran sequentially (e.g.: a connection should be started before the createChannel method is called) which would not be guaranteed if each method ran in a separate thread.

Reply:
Given this,  do you have any recommendations for a better way to solve this issue?